shader/main3d.frag.glsl

```diff
@@ -14,90 +17,153 @@ vec4 quantize4Bit(in vec4 color) {
+    default: return sampleSampler(tex0, tileConf, uvCoord, texFilter);
+    case 1: return sampleSampler(tex1, tileConf, uvCoord, texFilter);
+    case 2: return sampleSampler(tex2, tileConf, uvCoord, texFilter);
+    case 3: return sampleSampler(tex3, tileConf, uvCoord, texFilter);
+    case 4: return sampleSampler(tex4, tileConf, uvCoord, texFilter);
+    case 5: return sampleSampler(tex5, tileConf, uvCoord, texFilter);
+    case 6: return sampleSampler(tex6, tileConf, uvCoord, texFilter);
+    case 7: return sampleSampler(tex7, tileConf, uvCoord, texFilter);
+  }
+}
 
-  ivec2 lower_flag = ivec2(step(0.0, ratio.xz - ratio.yw));
-  ivec4 corner = ivec4(
-    lower_flag.x, 1 - lower_flag.x,
-    lower_flag.y, 1 - lower_flag.y
-  );
+float computeLOD(inout uint tileIndex0, inout uint tileIndex1) {
+  // https://github.com/rt64/rt64/blob/0ca92eeb6c2f58ce3581c65f87f7261b8ac0fea0/src/shaders/TextureSampler.hlsli#L18
+  if (textLOD() == G_TL_TILE)
+    return 1.0f;
+  const uint texDetail = textDetail();
+  const bool lodSharpen = texDetail == G_TD_SHARPEN;
+  const bool lodDetail = texDetail == G_TD_DETAIL;
+  const bool lodSharpDetail = lodSharpen || lodDetail;
+
+#ifdef GL_ARB_derivative_control
+  const vec2 dfd = abs(vec2(dFdxCoarse(inputUV.x), dFdyCoarse(inputUV.y)));
+#else
+  const vec2 dfd = abs(vec2(dFdx(inputUV.x), dFdy(inputUV.y)));
+#endif
+  float maxDst = max(dfd.x, dfd.y);
 
-  ivec4 uv1 = uv0 + corner;
-  ivec4 uv2 = uv0 + 1;
+  if (lodSharpDetail) 
+    maxDst = max(maxDst, material.primLod.y);
 
-  uv0 = wrappedMirror(texSize, uv0);
-  uv1 = wrappedMirror(texSize, uv1);
-  uv2 = wrappedMirror(texSize, uv2);
+  int tileBase = int(floor(log2(maxDst)));
+  float lodFraction = maxDst / pow(2, max(tileBase, 0)) - 1.0;
 
-  vec4 v0 = vec4(0 - corner);
-  vec4 v1 = vec4(1 - corner);
-  vec4 v2 = ratio - vec4(corner);
+  if (lodSharpen && maxDst < 1.0)
+    lodFraction = maxDst - 1.0;
 
-  vec2 den = v0.xw * v1.yz - v1.xw * v0.yz;
+  if (lodDetail) {
+    if (lodFraction < 0.0)
+      lodFraction = maxDst;
+    tileBase += 1;
+  } else if (tileBase >= material.mipCount)
+    lodFraction = 1.0;
 
-  vec2 lambda1 = abs((v2.xz * v1.yw - v1.xz * v2.yw) / den);
-  vec2 lambda2 = abs((v0.xz * v2.yw - v2.xz * v0.yw) / den);
-  vec2 lambda0 = 1.0 - lambda1 - lambda2;
+  if (lodSharpDetail) 
+    tileBase = max(tileBase, 0);
+  else 
+    lodFraction = max(lodFraction, 0.0);
 
-  texData0 =  quantizeTexture0(texelFetch(tex0, uv1.xy, 0)) * lambda0.x
-            + quantizeTexture0(texelFetch(tex0, uv0.xy, 0)) * lambda1.x
-            + quantizeTexture0(texelFetch(tex0, uv2.xy, 0)) * lambda2.x;
-  
-  texData1 =  quantizeTexture1(texelFetch(tex1, uv1.zw, 0)) * lambda0.y
-            + quantizeTexture1(texelFetch(tex1, uv0.zw, 0)) * lambda1.y
-            + quantizeTexture1(texelFetch(tex1, uv2.zw, 0)) * lambda2.y;
+  tileIndex0 = clamp(tileBase, 0, material.mipCount);
+  tileIndex1 = clamp(tileBase + 1, 0, material.mipCount);
```

Comment:
I've mentioned this in an RT64 issue as well (https://github.com/rt64/rt64/issues/116) but some of this clamping behavior is not fully accurate to hardware. I suspect it doesn't matter for previewing if you have ensured that max lod level > min lod level, but mentioning just in case.

Reply:
i think its better to merge for now but I'll look at it!